pkg/routeagent_driver/handlers/ovn/submarinerGWRouteHandler.go

```diff
@@ -0,0 +1,179 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type GatewayRouteHandler struct {
+	event.HandlerBase
+	mutex           sync.Mutex
+	smClient        submarinerClientset.Interface
+	config          *environment.Specification
+	remoteEndpoints map[string]*submarinerv1.Endpoint
+	isGateway       bool
+	nextHopIP       string
+}
+
+func NewGatewayRouteHandler(env *environment.Specification, smClientSet submarinerClientset.Interface) *GatewayRouteHandler {
+	// We'll panic if env is nil, this is intentional
+	return &GatewayRouteHandler{
+		config:          env,
+		smClient:        smClientSet,
+		remoteEndpoints: map[string]*submarinerv1.Endpoint{},
+	}
+}
+
+func (h *GatewayRouteHandler) Init() error {
+	logger.Info("Starting GatewayRouteHandler")
+
+	nextHopIP, err := getNextHopOnK8sMgmtIntf()
+	if err != nil || nextHopIP == "" {
+		return errors.Wrapf(err, "error getting the ovn kubernetes management interface IP")
+	}
+
+	h.nextHopIP = nextHopIP
+
+	return nil
+}
+
+func (h *GatewayRouteHandler) GetName() string {
+	return "submariner-gw-route-handler"
+}
+
+func (h *GatewayRouteHandler) GetNetworkPlugins() []string {
+	// TODO enable when we switch to new implementation
+	// return []string{cni.OVNKubernetes}
+	return []string{}
+}
+
+func (h *GatewayRouteHandler) RemoteEndpointCreated(endpoint *submarinerv1.Endpoint) error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	h.remoteEndpoints[endpoint.Name] = endpoint
+
+	if h.isGateway {
+		_, err := h.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).Create(context.TODO(),
+			h.newGatewayRoute(endpoint), metav1.CreateOptions{})
+		if err != nil && !apierrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, "error processing the remote endpoint creation for %q", endpoint.Name)
+		}
+	}
+
+	return nil
+}
+
+func (h *GatewayRouteHandler) RemoteEndpointRemoved(endpoint *submarinerv1.Endpoint) error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	delete(h.remoteEndpoints, endpoint.Name)
+
+	if h.isGateway {
+		if err := h.smClient.SubmarinerV1().GatewayRoutes(endpoint.Namespace).Delete(context.TODO(),
+			endpoint.Name, metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+			return errors.Wrapf(err, "error deleting gatewayRoute %q", endpoint.Name)
+		}
+	}
+
+	return nil
+}
+
+func (h *GatewayRouteHandler) TransitionToNonGateway() error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	h.isGateway = false
+
+	return nil
+}
+
+func (h *GatewayRouteHandler) TransitionToGateway() error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	h.isGateway = true
+
+	for _, endpoint := range h.remoteEndpoints {
+		gwr := h.newGatewayRoute(endpoint)
+
+		result, err := util.CreateOrUpdate(context.TODO(), h.gatewayResourceInterface(endpoint.Namespace), gwr,
+			func(existing runtime.Object) (runtime.Object, error) {
+				existingGwr := gwr
+				return existingGwr, nil
+			})
```

Comment:
The assignment to `existingGwr` is unnecessary but you can use `util.Replace` instead.

```suggestion
		result, err := util.CreateOrUpdate(context.TODO(), h.gatewayResourceInterface(endpoint.Namespace), gwr, util.Replace(gwr))
```

Reply:
Done.